File: 15-3sum/3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    nums.sort((a, b) => a - b)
    const res = []

    for (let i = 0; i < nums.length; i++) {
        let low = i + 1, high = nums.length - 1, sum = 0;

        while (low < high) {
            sum = nums[i] + nums[low] + nums[high]
            if (sum === 0) {
                res.push([nums[i], nums[low], nums[high]])
                while (nums[low] === nums[low+1]) low++
                while(nums[high] === nums[high-1]) high--
                low++
                high--
            } else if (sum < 0) low++
            else high--
        }
        while (nums[i+1] === nums[i]) i++
    }
    return res
};

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()

        for i, val in enumerate(nums):
            if i > 0 and val == nums[i-1]:
                continue

            l, r = i + 1, len(nums) - 1

            while l < r:
                total = val + nums[l] + nums[r]

               
                if total < 0:
                    l += 1
                elif total > 0:
                    r -= 1
                else:
                    res.append([val, nums[l], nums[r]])

                    l += 1
                    r -= 1

                    while l < r and nums[l] == nums[l-1]:
                        l += 1
                    while l < r and nums[r] == nums[r+1]:
                        r -= 1

        return res